src/inc/til/rectangle.h

```diff
@@ -0,0 +1,484 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "point.h"
+#include "size.h"
+#include "some.h"
+
+#ifdef UNIT_TESTING
+class RectangleTests;
+#endif
+
+namespace til // Terminal Implementation Library. Also: "Today I Learned"
+{
+    class rectangle
+    {
+    public:
+        constexpr rectangle() noexcept :
+            rectangle(til::point{ 0, 0 }, til::point{ 0, 0 })
+        {
+        }
+
+        // On 64-bit processors, int and ptrdiff_t are different fundamental types.
+        // On 32-bit processors, they're the same which makes this a double-definition
+        // with the `ptrdiff_t` one below.
+#if defined(_M_AMD64) || defined(_M_ARM64)
+        constexpr rectangle(int left, int top, int right, int bottom) noexcept :
+            rectangle(til::point{ left, top }, til::point{ right, bottom })
+        {
+        }
+#endif
+
+        rectangle(size_t left, size_t top, size_t right, size_t bottom) :
+            rectangle(til::point{ left, top }, til::point{ right, bottom })
+        {
+        }
+
+        constexpr rectangle(ptrdiff_t left, ptrdiff_t top, ptrdiff_t right, ptrdiff_t bottom) noexcept :
+            rectangle(til::point{ left, top }, til::point{ right, bottom })
+        {
+        }
+
+        // Creates a 1x1 rectangle with the given top-left corner.
+        rectangle(til::point topLeft) :
+            _topLeft(topLeft)
+        {
+            _bottomRight = _topLeft + til::point{ 1, 1 };
+        }
+
+        // Creates a rectangle where you specify the top-left corner (included)
+        // and the bottom-right corner (excluded)
+        constexpr rectangle(til::point topLeft, til::point bottomRight) noexcept :
+            _topLeft(topLeft),
+            _bottomRight(bottomRight)
+        {
+        }

```

Comment:
What if `topLeft` is not above or left of `bottomRight`? It'd be cool if we wouldn't have to think of that and the rectangle smartly assigns two points to the right ones.

Reply:
I don't love when things try to infer what i mean, i dunno

Reply:
The part that gets weird is, if bottomRight is meant to be exclusive, and topLeft is inclusive, getting the two mixed up can cause some REAL problems.

Reply:
I really don't think that we should support users passing in coordinates in the wrong order, _especially_ now that you've mentioned how poorly it interacts with exclusivity ;)

Reply:
@carlos-zamora, are you thinking about this in the context of selection where you just want whatever is the valid rectangle between the two points? 

I'd think we might come up with something explicit that handles that scenario then, not an implicit rearrangement inside this constructor.

Reply:
Fair enough. I know in the selection code, it'd be helpful to just always have an idea of which anchor is "before" the other. But I can just do that check when I need it.